File: android/app/src/main/java/tv/swarmcast/p2p/PeerReputation.kt
package tv.swarmcast.p2p

import java.util.concurrent.ConcurrentHashMap

enum class PeerReputationEvent {
    SUCCESS,
    TIMEOUT,
    REJECT,
    HASH_MISMATCH
}

data class PeerReputationSnapshot(
    val peerId: String,
    val score: Int,
    val successes: Int,
    val failures: Int,
    val poisonOffenses: Int,
    val disconnected: Boolean
)

class PeerReputation(
    private val peerId: String,
    private val maxPoisonOffenses: Int = 2,
    private val maxScore: Int = 100,
    private val minScore: Int = -100
) {
    private var score = 0
    private var successes = 0
    private var failures = 0
    private var poisonOffenses = 0
    private var disconnected = false

    @Synchronized
    fun record(event: PeerReputationEvent): PeerReputationSnapshot {
        when (event) {
            PeerReputationEvent.SUCCESS -> {
                successes += 1
                score = (score + 3).coerceAtMost(maxScore)
            }
            PeerReputationEvent.REJECT -> {
                failures += 1
                score = (score - 1).coerceAtLeast(minScore)
            }
            PeerReputationEvent.TIMEOUT -> {
                failures += 1
                score = (score - 3).coerceAtLeast(minScore)
            }
            PeerReputationEvent.HASH_MISMATCH -> {
                failures += 1
                poisonOffenses += 1
                score = (score - 25).coerceAtLeast(minScore)
                if (poisonOffenses >= maxPoisonOffenses) disconnected = true
            }
        }
        return snapshot()
    }

    @Synchronized
    fun snapshot(): PeerReputationSnapshot =
        PeerReputationSnapshot(
            peerId = peerId,
            score = score,
            successes = successes,
            failures = failures,
            poisonOffenses = poisonOffenses,
            disconnected = disconnected
        )
}

class PeerReputationBook(
    private val maxPoisonOffenses: Int = 2,
    private val maxScore: Int = 100,
    private val minScore: Int = -100
) {
    private val peers = ConcurrentHashMap<String, PeerReputation>()

    fun get(peerId: String): PeerReputation =
        peers.computeIfAbsent(peerId) {
            PeerReputation(
                peerId = it,
                maxPoisonOffenses = maxPoisonOffenses,
                maxScore = maxScore,
                minScore = minScore
            )
        }

    fun record(peerId: String, event: PeerReputationEvent): PeerReputationSnapshot =
        get(peerId).record(event)

    fun score(peerId: String): Int = get(peerId).snapshot().score

    fun isDisconnected(peerId: String): Boolean = get(peerId).snapshot().disconnected

    fun candidates(): List<PeerReputationSnapshot> =
        peers.values
            .map { it.snapshot() }
            .filterNot { it.disconnected }
            .sortedByDescending { it.score }
}

File: packages/p2p/src/peerReputation.js
export const PEER_EVENT = Object.freeze({
  SUCCESS: "success",
  TIMEOUT: "timeout",
  REJECT: "reject",
  HASH_MISMATCH: "hash_mismatch"
});

export class PeerReputation {
  constructor({
    peerId,
    maxPoisonOffenses = 2,
    maxScore = 100,
    minScore = -100
  }) {
    this.peerId = peerId;
    this.maxPoisonOffenses = maxPoisonOffenses;
    this.maxScore = maxScore;
    this.minScore = minScore;
    this.score = 0;
    this.successes = 0;
    this.failures = 0;
    this.poisonOffenses = 0;
    this.disconnected = false;
  }

  record(event) {
    switch (event) {
      case PEER_EVENT.SUCCESS:
        this.successes += 1;
        this.score = Math.min(this.maxScore, this.score + 3);
        break;
      case PEER_EVENT.REJECT:
        this.failures += 1;
        this.score = Math.max(this.minScore, this.score - 1);
        break;
      case PEER_EVENT.TIMEOUT:
        this.failures += 1;
        this.score = Math.max(this.minScore, this.score - 3);
        break;
      case PEER_EVENT.HASH_MISMATCH:
        this.failures += 1;
        this.poisonOffenses += 1;
        this.score = Math.max(this.minScore, this.score - 25);
        if (this.poisonOffenses >= this.maxPoisonOffenses) this.disconnected = true;
        break;
      default:
        throw new Error(`unknown peer reputation event: ${event}`);
    }
    return this.snapshot();
  }

  snapshot() {
    return {
      peerId: this.peerId,
      score: this.score,
      successes: this.successes,
      failures: this.failures,
      poisonOffenses: this.poisonOffenses,
      disconnected: this.disconnected
    };
  }
}

export class PeerReputationBook {
  constructor(options = {}) {
    this.options = options;
    this.peers = new Map();
  }

  get(peerId) {
    if (!this.peers.has(peerId)) {
      this.peers.set(peerId, new PeerReputation({ peerId, ...this.options }));
    }
    return this.peers.get(peerId);
  }

  record(peerId, event) {
    return this.get(peerId).record(event);
  }

  candidates() {
    return [...this.peers.values()]
      .filter((peer) => !peer.disconnected)
      .sort((a, b) => b.score - a.score)
      .map((peer) => peer.snapshot());
  }
}
